src/controllers/sharedExecutionPlanUtils.ts

```diff
@@ -0,0 +1,184 @@
+            counter += 1;
+        }
+        filename = vscode.Uri.joinPath(folder, `plan${counter}.sqlplan`);
+    } else {
+        filename = vscode.Uri.joinPath(folder, "plan.sqlplan");
+    }
+
+    // Show a save dialog to the user
+    const saveUri = await vscode.window.showSaveDialog({
+        defaultUri: filename,
+        filters: {
+            "SQL Plan Files": ["sqlplan"],
+        },
+    });
+
+    if (saveUri) {
+        // Write the content to the new file
+        void vscode.workspace.fs.writeFile(
+            saveUri,
+            Buffer.from(payload.sqlPlanContent),
+        );
+    }
+
+    return state;
+}
+
+export async function showPlanXml(
+    state: QueryResultWebviewState | ExecutionPlanWebviewState,
+    payload: ExecutionPlanReducers["showPlanXml"],
+) {
+    const planXmlDoc = await vscode.workspace.openTextDocument({
+        content: formatXml(payload.sqlPlanContent),
+        language: "xml",
+    });
+
+    void vscode.window.showTextDocument(planXmlDoc);
+
+    return state;
+}
+
+export async function showQuery(
+    state: QueryResultWebviewState | ExecutionPlanWebviewState,
+    payload: ExecutionPlanReducers["showQuery"],
+    untitledSqlDocumentService: UntitledSqlDocumentService,
+) {
+    void untitledSqlDocumentService.newQuery(payload.query);
+
+    return state;
+}
+
+export async function updateTotalCost(
+    state: QueryResultWebviewState | ExecutionPlanWebviewState,
+    payload: ExecutionPlanReducers["updateTotalCost"],
+) {
+    return {
+        ...state,
+        executionPlanState: {
+            ...state.executionPlanState,
+            totalCost: (state.executionPlanState.totalCost +=
+                payload.addedCost),
+        },
+    };
+}
+
+export async function createExecutionPlanGraphs(
+    state: QueryResultWebviewState | ExecutionPlanWebviewState,
+    executionPlanService: ExecutionPlanService,
+    xmlPlans: string[],
+) {
+    let newState = {
+        ...state.executionPlanState,
+    };
+    const startTime = performance.now(); // timer for telemetry
+    for (const plan of xmlPlans) {
+        const planFile: ExecutionPlanGraphInfo = {
+            graphFileContent: plan,
+            graphFileType: ".sqlplan",
+        };
+        try {
+            newState.executionPlanGraphs = newState.executionPlanGraphs.concat(
+                (await executionPlanService.getExecutionPlan(planFile)).graphs,
+            );
+
+            newState.loadState = ApiStatus.Loaded;
+
+            sendActionEvent(
+                TelemetryViews.QueryPlan,
+                TelemetryActions.OpenQueryPlan,
+                {},
+                {
+                    numberOfPlans:
+                        state.executionPlanState.executionPlanGraphs.length,
+                    LoadTimeInMs: performance.now() - startTime,
+                },
+            );
+        } catch (e) {
+            newState.loadState = ApiStatus.Error;
+            newState.errorMessage = e.toString();

```

Comment:
use `getErrorMessage(err)` since anything can be thrown as an error, and this function will handle more cases cleanly.